src/schema/v2/artwork/collectorSignals.ts

```diff
@@ -0,0 +1,95 @@
+import { GraphQLFieldConfig } from "graphql"
+import { GraphQLInt, GraphQLObjectType } from "graphql"
+import { ResolverContext } from "types/graphql"
+import { PartnerOfferToCollectorType } from "../partnerOfferToCollector"
+
+export const collectorSignals: GraphQLFieldConfig<any, ResolverContext> = {
+  type: new GraphQLObjectType({
+    description: "Collector signals available to the artwork",
+    name: "CollectorSignals",
+    fields: {
+      bidCount: {
+        type: GraphQLInt,
+        description: "Lot bid count",
+      },
+      lotWatcherCount: {
+        type: GraphQLInt,
+        description: "Lot watcher count",
+      },
+      partnerOffer: {
+        type: PartnerOfferToCollectorType,
+        description: "Partner offer available to collector",
+      },
+    },
+  }),
+  description: "List of collector signals",
+
+  resolve: async (artwork, {}, _ctx) => {
+    return artwork.collectorSignals
+  },
+}
+
+interface EnrichedSignals {
+  bidCount?: number
+  lotWatcherCount?: number
+  partnerOffer?: { endAt: string }
+}
+
+export const enrichArtworkWithCollectorSignals = async (artwork, ctx) => {
+  const enrichedSignals: EnrichedSignals = {}
+  artwork.collectorSignals = enrichedSignals
+
+  const artworkId = artwork._id
+
+  const inSale = artwork.sale_ids?.length > 0
+
+  // Handle signals for auction artworks
+  if (inSale) {
+    const activeAuctions = await activeAuctionsForArtwork(artwork, ctx)
+    const activeAuction = activeAuctions[0]
+
+    if (activeAuction) {
+      const saleArtworkRequest = ctx.saleArtworkLoader({
+        saleId: activeAuction.id, // _id?
+        saleArtworkId: artworkId,
+      })
+      const [activeSaleArtwork] = await Promise.all([saleArtworkRequest])
+
+      if (activeSaleArtwork) {
+        if (artwork.recent_saves_count) {
+          enrichedSignals.lotWatcherCount = artwork.recent_saves_count
+        }
+        if (activeSaleArtwork.bidder_positions_count) {
+          enrichedSignals.bidCount = activeSaleArtwork.bidder_positions_count
+        }
+      }
+    }
+
+    // Handle signals for non-auction acquirable artworks
+  } else if (artwork.acquireable) {
+    if (ctx.mePartnerOffersLoader) {
+      const partnerOffer = await ctx
+        .mePartnerOffersLoader({
+          artwork_id: artworkId,
+          sort: "-created_at",
+          size: 1,
+        })
+        .then(({ body }) => body[0])
+      enrichedSignals.partnerOffer = partnerOffer
+    }
+  }
+  return artwork
+}
+
+const activeAuctionsForArtwork = async (artwork, ctx) => {
+  if (artwork.sale_ids?.length > 0) {
+    const { salesLoader } = ctx
+    const sales = await salesLoader({
+      id: artwork.sale_ids,
+      is_auction: true,
+      live: true,
+    })
+    return sales
+  }
+  return []
+}
```

Comment:
(I know this is still wip, but you asked for feedback in slack :)

Thinking: A few of the patterns in here feel a bit idiosyncratic compared to the rest of MP, but maybe i'm not understanding exactly why a lot of this is necessary. Would you mind kinda breaking down in the PR description what the flow is architecture-wise in a bit more detail?

Reply:
For sure - the auction signals logic was the hardest part. I tried to use the [saleArtworksAllLoader](https://github.com/artsy/metaphysics/blob/613ac7be7c685131905184beaee37134ee678759/src/lib/loaders/loaders_without_authentication/gravity.ts#L223) here but ran into some auth issues. Was hoping i could request `/sale_artworks` with `artwork_id` and `biddable_sale: true`. So what i am doing instead is

1. testing if the artwork has `sale_ids`, otherwise skipping
2. querying for those sale ids + `is_auction` + `live`, which follows the logic in [Artwork.biddable](https://github.com/artsy/metaphysics/blob/50c94eb6a3f1ceb1a8223041bbb15b6b5fdf7fbd/src/schema/v2/artwork/index.ts#L745-L761).
3. Then up above, below where this function is called, if we got a biddable auction, fetch the sale artwork for this sale/artwork combination [as we do in Artwork.saleArtwork](https://github.com/artsy/metaphysics/blob/50c94eb6a3f1ceb1a8223041bbb15b6b5fdf7fbd/src/schema/v2/artwork/index.ts#L1508-L1528)). If we get that back then we know it is visible to the user and we can use that for bid count.
4. For lot watcher count I re-used the `artwork.recent_saves_count` though we might want to get more specific in the future.

There are two things I could do to try and clean this up:
- make that first approach work, maybe i was missing some minor detail
- Skip 2 up above

Reply:
This makes sense! Ty. Might be good to drop it in the PR description 👍